File: cpp/binary-trees/binary-trees.cpp
#include <iostream>

struct TreeNode {
    TreeNode* left;
    TreeNode* right;

    TreeNode() : left(nullptr), right(nullptr) {}
    ~TreeNode() {
        delete left;
        delete right;
    }
};

TreeNode* make_tree(int depth) {
    if (depth > 0) {
        depth--;
        TreeNode* node = new TreeNode();
        node->left = make_tree(depth);
        node->right = make_tree(depth);
        return node;
    }
    return nullptr;
}

int check_tree(TreeNode* node) {
    if (node == nullptr) {
        return 0;
    }
    return 1 + check_tree(node->left) + check_tree(node->right);
}

int main() {
    for (int i = 0; i < 45; i++) {
        int depth = 45;
        TreeNode* tree = make_tree(depth);
        int total_nodes = check_tree(tree);
        std::cout << "Tree of depth " << depth << " has " << total_nodes << " nodes." << std::endl;
        delete tree;
    }
    return 0;
}

File: ruby/binary-trees/binary-trees.rb
class TreeNode
    attr_accessor :left, :right
  
    def initialize
      @left = nil
      @right = nil
    end
  end
  
  def make_tree(depth)
    return nil if depth <= 0
    node = TreeNode.new
    depth -= 1
    node.left = make_tree(depth)
    node.right = make_tree(depth)
    node
  end
  
  def check_tree(node)
    return 0 unless node
    1 + check_tree(node.left) + check_tree(node.right)
  end
  
  (0..45).each do |i|
    depth = 45
    tree = make_tree(depth)
    total_nodes = check_tree(tree)
    puts "Tree of depth #{depth} has #{total_nodes} nodes."
  end
  